### ISSUE TITLE
Interceptors do not report nullability warnings when type parameter constraints are not satisfied

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#70310]
**Version Used**: 
```
Microsoft Visual Studio Professional 2022
Version 17.8.0 Preview 2.0
VisualStudio.17.Preview/17.8.0-pre.2.0+34112.27
Microsoft .NET Framework
Version 4.8.09037
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Runtime.CompilerServices;

partial class C
{
  public string? Method1<T>(T arg) => null;
}

static class Program
{
  public static void Main()
  {
    var c = new C();
    string? x = null;

    c.Method1(x);
  }
}

partial class C
{
  [InterceptsLocation("Program.cs", 16, 7)]
  public string? Interceptor1<T>(T arg) where T: notnull => arg.ToString();
}

namespace System.Runtime.CompilerServices
{
  [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
  public class InterceptsLocationAttribute : Attribute
  {
    public InterceptsLocationAttribute(string path, int line, int character) { }
  }
}
```

**Expected Behavior**:
A warning is reported for the intercepted call `c.Method1(x)` which is translated into `C.Interceptor1<string?>(x);` because the type argument doesn't satisfy the `where T: notnull` constraint on the interceptor method

**Actual Behavior**:
No warnings at all. The code crashes at runtime with a `NullReferenceException`

**Comment by RikkiGibson** [comment]
The problem here is that the interceptor's type parameter constraints are checked during lowering. At that point, the nullable-reinferred version of the call has been discarded, and we are lowering the initially-bound version of the call.

https://github.com/dotnet/roslyn/blob/66481154d73696a7a4591e2b9efecd679efead29/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs#L177

It's not straightforward to solve, unfortunately, and I think I'd want to pursue some of the longer-term design questions before investing effort into solving it.
